apollo-router/src/query_planner/tests.rs

```diff
@@ -985,3 +985,489 @@ async fn missing_fields_in_requires() {
+{
+id: ID!
+}
+
+type Query
+@join__type(graph: AUTHOR_SUBGRAPH)
+@join__type(graph: BOOK_SUBGRAPH)
+@join__type(graph: NODE_RELAY_SUBGRAPH)
+{
+b: Boolean @join__field(graph: AUTHOR_SUBGRAPH)
+book: Book @join__field(graph: BOOK_SUBGRAPH)
+node(id: ID): Node @join__field(graph: NODE_RELAY_SUBGRAPH)
+}"#;
+
+#[tokio::test]
+async fn typename_propagation() {
+    let subgraphs = MockedSubgraphs(
+        [
+            ("author_subgraph", MockSubgraph::builder().with_json(
+                serde_json::json! {{
+                    "query": "query QueryBook__author_subgraph__1($representations:[_Any!]!){_entities(representations:$representations){...on Author{fullName}}}",
+                    "operationName": "QueryBook__author_subgraph__1",
+                    "variables": {
+                        "representations": [{
+                            "__typename": "Author",
+                            "authorId": "Author1"
+                        }]
+                    }
+                }},
+                serde_json::json! {{"data": {
+                    "_entities": [{
+                        "fullName": "Ada"
+                    }]
+                } }},
+            ).build()),
+            ("book_subgraph", MockSubgraph::builder().build()),
+            ("node_relay_subgraph", MockSubgraph::builder().with_json(
+                serde_json::json! {{
+                    "query": "query Query__node_relay_subgraph__0{node{__typename ...on Book{id author{__typename}}}}",
+                    "operationName": "Query__node_relay_subgraph__0"
+                }},
+                serde_json::json! {{"data": {
+                    "node": {
+                      "__typename": "Book",
+                      "id": "1",
+                      "author": {
+                        "__typename": "Author"
+                      }
+                    }
+                } }},
+            ).build()),
+        ]
+        .into_iter()
+        .collect(),
+    );
+
+    let service = TestHarness::builder()
+        .configuration_json(serde_json::json!({"include_subgraph_errors": { "all": true } }))
+        .unwrap()
+        .schema(TYPENAME_PROPAGATION_SCHEMA)
+        .extra_plugin(subgraphs)
+        .build_supergraph()
+        .await
+        .unwrap();
+
+    let request = supergraph::Request::fake_builder()
+        .context(Context::new())
+        .query(
+            "query Query {
+            node {
+              __typename
+              ... on Book {
+                id
+                author {
+                  __typename
+                }
+              }
+            }
+          }",
+        )
+        .build()
+        .unwrap();
+
+    let mut stream = service.clone().oneshot(request).await.unwrap();
+    let response = stream.next_response().await.unwrap();
+    insta::assert_json_snapshot!(serde_json::to_value(&response).unwrap());
+}
+
+#[tokio::test]
+async fn typename_propagation2() {
```

Comment:
this test is still failing. I do not know yet if we should fix it (it requires some work to find out the right type) or accept that the implementation does what it can, and won't cover all cases (missing `__typename` can generally be fixed by adding it to the `@requires` selection)